src/LanguageServer/Protocol/Handler/Completion/CompletionResolveHandler.cs

```diff
@@ -23,59 +25,94 @@ namespace Microsoft.CodeAnalysis.LanguageServer.Handler
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CompletionResolveHandler(IGlobalOptionService globalOptions)
         {
             _globalOptions = globalOptions;
-            _completionListCache = completionListCache;
         }
 
         public LSP.TextDocumentIdentifier? GetTextDocumentIdentifier(LSP.CompletionItem request)
-            => CompletionResolveHandler.GetTextDocumentCacheEntry(request);
+            => GetTextDocumentCacheEntry(request);
 
-        public async Task<LSP.CompletionItem> HandleRequestAsync(LSP.CompletionItem completionItem, RequestContext context, CancellationToken cancellationToken)
+        public Task<LSP.CompletionItem> HandleRequestAsync(LSP.CompletionItem completionItem, RequestContext context, CancellationToken cancellationToken)
         {
-            var cacheEntry = GetCompletionListCacheEntry(completionItem);
-            if (cacheEntry == null)
+            var completionListCache = context.GetRequiredLspService<CompletionListCache>();
+
+            if (!completionListCache.TryGetCompletionListCacheEntry(completionItem, out var cacheEntry))
             {
                 // Don't have a cache associated with this completion item, cannot resolve.
                 context.TraceInformation("No cache entry found for the provided completion item at resolve time.");
-                return completionItem;
+                return Task.FromResult(completionItem);
             }
 
             var document = context.GetRequiredDocument();
-            var completionService = document.Project.Services.GetRequiredService<CompletionService>();
-
-            // Find the matching completion item in the completion list
-            var selectedItem = cacheEntry.CompletionList.ItemsList.FirstOrDefault(cachedCompletionItem => MatchesLSPCompletionItem(completionItem, cachedCompletionItem));
-
             var completionOptions = _globalOptions.GetCompletionOptions(document.Project.Language);
             var symbolDescriptionOptions = _globalOptions.GetSymbolDescriptionOptions(document.Project.Language);
+            var capabilityHelper = new CompletionCapabilityHelper(context.GetRequiredClientCapabilities());
+
+            return ResolveCompletionItemAsync(
+                completionItem, cacheEntry.CompletionList, document, completionOptions, symbolDescriptionOptions, capabilityHelper, cancellationToken);
+        }
+
+        public static Task<LSP.CompletionItem> ResolveCompletionItemAsync(
+            LSP.CompletionItem completionItem,
+            Document document,
+            CompletionOptions completionOptions,
+            SymbolDescriptionOptions symbolDescriptionOptions,
+            CompletionCapabilityHelper capabilityHelper,
+            CompletionListCache completionListCache,
+            CancellationToken cancellationToken)
+        {
+            if (!completionListCache.TryGetCompletionListCacheEntry(completionItem, out var cacheEntry))
```

Comment:
Is it possible to move retrieving the cache entry into the shared overload?  Instead of doing it once here and also in HandleRequestAsync

that path should rarely be hit anyway, so its fine to do a small bit of unnecessary work before it.

Reply:
The reason I didn't do this was to avoid a behavior change. `HandleRequestAsync` logs a message using the `RequestContext` on a cache miss. If you're OK with that message being removed, I'm happy to make the change.

Reply:
Followed up with @dibarbet offline, and they're OK with not unifying further to leave the logging message.